luatest/treegen.lua

```diff
@@ -0,0 +1,210 @@
+
+local fio = require('fio')
+local fun = require('fun')
+local checks = require('checks')
+
+local log = require('luatest.log')
+
+local treegen = {}
+
+local function find_template(group, script)
+    for _, template_def in ipairs(group.templates) do
+        if script:match(template_def.pattern) then
+            return template_def.template
+        end
+    end
+    error(("treegen: can't find a template for script %q"):format(script))
+end
+
+--- Write provided script into the given directory.
+--
+-- @string directory Directory where the script will be created.
+-- @string filename File to write (possible nested path: /foo/bar/main.lua).
+-- @string script The body of the script to write.
+-- @return string
+function treegen.write_script(directory, filename, script)
+    checks('string', 'string', 'string')
+    local script_abspath = fio.pathjoin(directory, filename)
+    local flags = {'O_CREAT', 'O_WRONLY', 'O_TRUNC'}
+    local mode = tonumber('644', 8)
+
+    local scriptdir_abspath = fio.dirname(script_abspath)
+    log.info('Creating a directory: %s', scriptdir_abspath)
+    fio.mktree(scriptdir_abspath)
+
+    log.info('Writing a script: %s', script_abspath)
+    local fh = fio.open(script_abspath, flags, mode)
+    fh:write(script)
+    fh:close()
+    return script_abspath
+end
+
+-- Generate a script that follows a template and write it at the
+-- given path in the given directory.
+--
+-- @table group Group of tests.
+-- @string directory Directory where the script will be created.
+-- @string filename File to write (possible nested path: /foo/bar/main.lua).
+-- @table replacements List of replacement templates.
+-- @return string
+local function gen_script(group, directory, filename, replacements)
+    checks('table', 'string', 'string', 'table')
+    local template = find_template(group, filename)
+    replacements = fun.chain({filename = filename}, replacements):tomap()
+    local body = template:gsub('<(.-)>', replacements)
+    return treegen.write_script(directory, filename, body)
+end
+
+--- Initialize treegen module in the given group of tests.
+--
+-- @tab group Group of tests.
+function treegen.init(group)
+    checks('table')
+    group.tempdirs = {}
+    group.templates = {}
+end
+
+--- Remove all temporary directories created by the test
+-- unless KEEP_DATA environment variable is set to a
+-- non-empty value.
+--
+-- @tab group Group of tests.
+function treegen.clean(group)
+    checks('table')
+    local dirs = table.copy(group.tempdirs) or {}
+    group.tempdirs = nil
+
+    local keep_data = (os.getenv('KEEP_DATA') or '') ~= ''
+
+    for _, dir in ipairs(dirs) do
+        if keep_data then
+            log.info('Left intact due to KEEP_DATA env var: %s', dir)
+        else
+            log.info('Recursively removing: %s', dir)
+            fio.rmtree(dir)
+        end
+    end
+
+    group.templates = nil
+end
+
+--- Save the template in the given group of test with the given pattern.
```

Comment:
Where the template is saved in an implementation detail.

Let's describe an effect of this function, not how it is implemented.

Reply:
I have shortened the description.